### PR TITLE
Add information about message direction to metamodel

### LINKED ISSUE #998: metamodel: Add information about whether a request/notification is sent from the client or the server
This would be handy for implementations which treat these kinds of messages differently, e.g. the Haskell implementation uses this to make sure servers don't accidentally register handlers for messages that can't be sent to them.

### PR DESCRIPTION
**Pull request by michaelpj**
Fixes #998.

It would be good to get the info checked! I went through and everything that said it's sent from the server got `serverToClient`, `progress` and `cancelRequest` got `both`, and everything else got `clientToServer`.

### CODE REVIEW

**Comment by michaelpj** [comment]
Okay, so the build fails because it can't build the visitor that auto-generates the metamodel, which isn't surprising since I'm not sure we *have* the information in the typescript implementation to use. This is really something we need to specify in the metamodel... Not sure what to do about that.

**Comment by dbaeumer** [comment]
@michaelpj sorry for not mentioning that earlier but we can't add this to the metaModel.json since that one is generated using `npm run generate:metaModel`. We nee to annotate the Request definitions. So we would do something like this:

```
export namespace CompletionRequest {
        export const direction: MessageDirection = MessageDirection.clientToServer;
	export const method: 'textDocument/completion' = 'textDocument/completion';
	export const type = new ProtocolRequestType<CompletionParams, CompletionItem[] | CompletionList | null, CompletionItem[], void, CompletionRegistrationOptions>(method);
}
```

and then adopt the generator to honor this information. I can help with fixing the generator if you do a PR that adds the information to the request definitions.

**Comment by michaelpj** [comment]
Okay, I think I've done that. I haven't touched the visitor or the generated metaModel. I can try and give that a shot, but I haven't actually managed to properly build the project yet...

**Comment by dbaeumer** [comment]
To build the project:

- clone it
- run npm install
- run npm run symlink
- run npm run compile

**Comment by dbaeumer** [comment]
Let me know if you need further help.

**Comment by michaelpj** [comment]
Okay, I built the project and all the protocol part compiles, so it looks like I did all the adding of `messageDirection` fields okay. But I'm not a typescript dev and it's not obvious to me how to modify the visitor. Could you take a look? I suspect it's just a couple of lines... I just don't know what they are!

The branch here has my work so far.

**Comment by michaelpj** [comment]
Done.

**Review by dbaeumer — DISMISSED** [review]

File: protocol/src/common/protocol.ts [hunk 1]
```diff
@@ -332,7 +332,9 @@ export interface RegistrationParams {
  * handler on the client side.
  */
 export namespace RegistrationRequest {
-	export const type = new ProtocolRequestType<RegistrationParams, void, never, void, void>('client/registerCapability');
+  export const method: 'client/registerCapability' = 'client/registerCapability';
```
Comment (aeschli) [inline protocol/src/common/protocol.ts:335]: Looks like there's an unusual indent

**Review by aeschli — APPROVED** [review]

**Comment by michaelpj** [comment]
@dbaeumer the build fails because the visitor isn't filling in the new field, but that requires fixing the visitor :( Feel free to push changes to my branch if you work on this!

File: protocol/src/common/protocol.ts (hunk 1 above)
Comment (michaelpj) [reply protocol/src/common/protocol.ts]: Fixed!

**Comment by dbaeumer** [comment]
@michaelpj even if I make `messageDirection` optional in the meta model (which we should do to get the PR in a state were we can merge it) I still have compile errors in your repository. The first one is:

capture

which is actually correct :-)

**Comment by dbaeumer** [comment]
`MessageDirection` should be defined on the protocol level and not the types level.

**Comment by michaelpj** [comment]
Thanks, I'll take another look, I thought I was down to just the visitor :)

Should I make it optional in the meta model for now? I guess that would make things compile and then we can make it mandatory later.

**Comment by michaelpj** [comment]
Alright, now it's green. I made `messageDirection` optional, but I think you should be able to make it mandatory when you update the visitor.

**Comment by dbaeumer** [comment]
Yes, that is the right way to go. We keep it optional for now until the visitor is updated.

File: protocol/src/common/protocol.$.ts [hunk 2]
```diff
@@ -15,12 +15,16 @@ import { WorkDoneProgressBegin, WorkDoneProgressEnd, WorkDoneProgressReport } fr
 
 // @ts-ignore 6196
 namespace SetTraceNotification {
-	export const type = new ProtocolNotificationType<SetTraceParams, void>('$/setTrace');
+  export const method: '$/setTrace' = '$/setTrace';
```
Comment (dbaeumer) [inline protocol/src/common/protocol.$.ts:18]: @michaelpj these indents still seem to be off. Looks like that use 2 spaces, but in this repository we indent using tabs. Could you please address this as well. 

Thanks.

**Review by dbaeumer — CHANGES_REQUESTED** [review]

File: protocol/src/common/protocol.$.ts (hunk 2 above)
Comment (michaelpj) [reply protocol/src/common/protocol.$.ts:18]: Argh sorry, I thought I'd got all of them >:(

**Pull request merged**